x/leverage/keeper/keeper.go

```diff
@@ -95,115 +96,58 @@ func (k Keeper) WithdrawAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, uToken
 		return sdkerrors.Wrap(types.ErrInsufficientBalance, uToken.String())
 	}
 
-	// ensure the tokens exist in the leverage module account's balance
-	// TODO: Use exchange rate instead of 1:1 redeeming
-	tokenDenom := k.FromUTokenToTokenDenom(ctx, uToken.Denom)
-	token := sdk.NewCoin(tokenDenom, uToken.Amount)
-	if !k.bankKeeper.HasBalance(ctx, authtypes.NewModuleAddress(types.ModuleName), token) {
-		// TODO: We should never enter this case -- consider a panic
-		return sdkerrors.Wrap(types.ErrInsufficientBalance, token.String())
+	// TODO: Calculate lender's borrow limit and borrowed value, if any, to prevent
+	// borrowers from withdrawing assets that are being used as collateral.
+
+	withdrawal, err := k.ExchangeUTokens(ctx, uToken)
+	if err != nil {
+		return err
+	}
+
+	// Ensure module account has sufficient unreserved tokens to withdraw
+	reservedAmount := k.GetReserveAmount(ctx, withdrawal.Denom)
+	availableAmount := k.bankKeeper.GetBalance(ctx, authtypes.NewModuleAddress(types.ModuleName), withdrawal.Denom).Amount
+	if withdrawal.Amount.GT(availableAmount.Sub(reservedAmount)) {
```

Comment:
Do we need to worry about negative values here?

Reply:
BankKeeper.GetBalance should be fine, but I'll put a guard on GetReserveAmount to make sure it never returns a negative value. 

(In practice it should be fine, as reserves only have `Get` and `Increase` functions, so achieving a negative value would have to occur by int overflow or a marshaling problem)